File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.ts
function queryResults(limit: number, queries: number[][]): number[] {
    const keyToValue = new Map<number, number>();  // Stores key -> value
    const valueCount = new Map<number, number>();  // Tracks occurrences of values
    const result: number[] = [];
    let distinctValues = 0;

    for (const [key, value] of queries) {
        // If key already exists, decrease count of old value
        if (keyToValue.has(key)) {
            const oldValue = keyToValue.get(key)!;
            valueCount.set(oldValue, (valueCount.get(oldValue) ?? 0) - 1);
            if (valueCount.get(oldValue) === 0) {
                distinctValues--; // Remove old value if count reaches 0
            }
        }

        // Assign new value to key
        keyToValue.set(key, value);
        if (!valueCount.has(value) || valueCount.get(value) === 0) {
            distinctValues++; // New distinct value added
        }
        valueCount.set(value, (valueCount.get(value) ?? 0) + 1);

        result.push(distinctValues);
    }

    return result;
}

File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.py
from collections import defaultdict

class Solution:
    def queryResults(self, limit: int, queries: list[list[int]]) -> list[int]:
        key_to_value = {}  # Stores key -> value
        value_count = defaultdict(int)  # Tracks occurrences of values
        result = []
        distinct_values = 0

        for key, value in queries:
            # If key already exists, decrease count of old value
            if key in key_to_value:
                old_value = key_to_value[key]
                value_count[old_value] -= 1
                if value_count[old_value] == 0:
                    distinct_values -= 1  # Remove old value if count reaches 0

            # Assign new value to key
            key_to_value[key] = value
            if value_count[value] == 0:
                distinct_values += 1  # New distinct value added
            value_count[value] += 1

            result.append(distinct_values)

        return result
